### ISSUE TITLE
Possible bug with fori_loop for jaxlib==0.1.60, jax==0.2.9 

### ISSUE THREAD

**Issue by gpescia** [google/jax#5779]
Hi,

I have currently updated my jaxlib from 0.1.55 to 0.1.60 as well as my jax from 0.2.1 to 0.2.9. 
Now I am getting different results for a method which includes a fori_loop.  

The minimal code to reproduce my issue is:
```python
import jax
import jax.numpy as np
from functools import partial

n = 64
L = 10.

key = jax.random.PRNGKey(0)
key, subkey = jax.random.split(key)
x = jax.random.uniform(key, shape=(3*n,), minval = 0, maxval = L).reshape(1,3*n,1)

@partial(jax.jit, static_argnums=(1,))
def minimum_length_v(x,n):
    s_dim = x.shape[0] // n
    y = x.reshape(-1,s_dim)
    
    temp = np.hstack(n*[y]).reshape(n,s_dim,n,order='F')
    temp = temp - np.transpose(y)[np.newaxis:,:]
    distances = np.abs(np.transpose(temp,(0,2,1))[np.triu_indices(n,1)])
    
    return np.linalg.norm(distances,axis=1)

def p_x(x,L):
    d = minimum_length_v(x,x.shape[0]//3)
    return np.sum(1./d**5)

dp_x = jax.jit(jax.vmap(jax.vmap(jax.grad(p_x),in_axes=(0,None),out_axes=0),\
                                    in_axes=(2,None),out_axes=2))

def test(x,n):
        unity = np.eye(x.shape[1])
        
        def second_deriv(i,state):
            save, temp = state
            unit = temp[jax.ops.index[i,:]].reshape(1,x.shape[1])
            d2p = jax.jit(jax.vmap(jax.vmap(jax.grad(lambda x,a: np.vdot(unit,jax.grad(p_x)(x,a))),\
                                            in_axes=(0,None),out_axes=0)\
                   ,in_axes=(2,None),out_axes=2))
            save = jax.ops.index_update(save,jax.ops.index[:,i,:],d2p(x,n)[:,i,:])
            return save, temp
        
        save = np.zeros((x.shape))
        save, unity = jax.lax.fori_loop(0,x.shape[1],second_deriv,(save,unity))
        """
        save = np.zeros((x.shape))
        for i in range(x.shape[1]):
            unit = unity[jax.ops.index[i,:]].reshape(1,x.shape[1])
            d2p = jax.jit(jax.vmap(jax.vmap(jax.grad(lambda x,a: np.vdot(unit,jax.grad(p_x)(x,a))),\
                                            in_axes=(0,None),out_axes=0)\
                   ,in_axes=(2,None),out_axes=2))
            save = jax.ops.index_update(save,jax.ops.index[:,i,:],d2p(x,n)[:,i,:])
        """
        return -0.5*np.sum((dp_x(x,n)/2)**2 + save/2,axis=1)

print(test(x,n))
```
Explanation: 
I have a function p_x(x) and samples of the form (n_samples, n_dims, n_chains) coming from a Metropolis random walk. 
I want to compute <img src="https://render.githubusercontent.com/render/math?math=-0.5 \sum_i \left( (\frac{d}{dx_i}p_x)^2 %2B \frac{d^2}{dx_i^2}p_x(x)\right) "> for all samples in each chain (the sum over  i goes over n_dims). 

In the file appended I get -1171920 for jaxlib==0.1.55 and jax==0.2.1 whereas for jaxlib==0.1.60 and jax==0.2.9 the output is -1171913.1. When I replace the fori_loop with a standard python for-loop, both versions yield -1171920 (indicating that the problem might be with the jaxlib==0.1.60, jax==0.2.9 version).

**Comment by hawkinsp** [comment]
I think this sort of relatively small difference is most likely due to floating-point reordering. Slight changes to the compiler and to JAX can have noticeable numerical effects, and this is legal and within specification.

If I enable float64 mode in JAX, by making the following changes:
```
jax.config.update("jax_enable_x64", True)
...
x = jax.random.uniform(key, shape=(3*n,), minval = 0, maxval = L, dtype=np.float32).reshape(1,3*n,1).astype(np.float64)
```

i.e., enabling x64 mode in the configuration, generating `float32` random numbers to make sure we get the same input, and then upcasting to `float64`, I get:
```
[[-1171912.16903163]]
```
which is closer to `-1171913.1`.

Do you have any reason to think the new answer is particularly inaccurate? If you do care about accuracy deeply, I'd suggest using `float64`.

One other thing I'll note is that code like this:
```
        for i in range(x.shape[1]):
            unit = unity[jax.ops.index[i,:]].reshape(1,x.shape[1])
            d2p = jax.jit(jax.vmap(jax.vmap(jax.grad(lambda x,a: np.vdot(unit,jax.grad(p_x)(x,a))),\
                                            in_axes=(0,None),out_axes=0)\
                   ,in_axes=(2,None),out_axes=2))
            save = jax.ops.index_update(save,jax.ops.index[:,i,:],d2p(x,n)[:,i,:])
```
is much better written in JAX using vectorization instead of an explicit `for` loop. e.g., Could you just `vmap` the whole thing? Unless I'm missing something there's no dependence between iterations of the loop.

**Comment by gpescia** [comment]
Thanks for answering!
In my application the difference is bigger than what is shown in the example above:
```import jax.numpy as np
import jax
​
from functools import partial
​
# compute pairwise distances with minimal image convention
@partial(jax.jit, static_argnums=(1,))
def minimum_length_v(x, n_particles, L):
    s_dim = x.shape[0] // n_particles
    y = x.reshape(-1, s_dim)
​
    temp = np.hstack(n_particles * [y]).reshape(n_particles, s_dim, n_particles, order='F')
    temp = temp - np.transpose(y)[np.newaxis:, :]
    distances = np.abs(np.transpose(temp, (0, 2, 1))[np.triu_indices(n_particles, 1)])
​
    return np.linalg.norm(np.where(np.abs(distances) > L / 2., L - distances, distances), axis=1)
​
​
def p_x(x, p, L):
    s_dim = 3
​
    d_v = minimum_length_v(x, x.shape[0] // s_dim, L)
​
    return np.sum(-((p[0][0][0] / d_v) ** 5))
​
​
n_particles = 50
​
rho = 1.96 * 10 ** (28)
rm = 2.9673 * 10 ** (-10)
​
L = (n_particles / (rho * rm ** 3)) ** (1 / 3)
​
x = np.array(
    [1.2714936, 2.3810654, 3.242951, 1.245318, 0.14467287, 0.5724811, 0.87262315, 4.009536, 1.5740701, 3.0824735,
     2.4342637, 0.503347, 2.962324, 3.4860227, 3.2341833, 0.3227515, 2.7460792, 1.8804604, 4.1680517, 3.7713027,
     0.5499155, 0.6238601, 2.595438, 0.64385766, 2.7458267, 1.8126817, 2.84528, 4.238092, 0.72629005, 1.6812574,
     0.37667072, 1.6731755, 2.8406723, 0.3504557, 3.0577617, 3.0469577, 2.7995348, 2.9948983, 1.7909292, 3.8149314,
     2.3564687, 1.8846831, 0.658647, 0.9721517, 0.013277531, 3.4270463, 0.782443, 2.6101816, 2.043499, 4.0885873,
     0.9977556, 3.388384, 3.2933455, 4.0780115, 3.6067586, 1.3991643, 0.8838493, 1.411048, 0.4048683, 2.8891149,
     2.5952082, 1.380391, 0.14428352, 2.049314, 4.0891585, 2.4663055, 3.7893374, 3.2627122, 2.802615, 0.44451624,
     4.0768976, 3.77346, 3.0581415, 1.7201387, 1.4119904, 1.5756953, 1.1075743, 2.405463, 3.04428, 1.5742427, 3.971295,
     1.6742224, 0.10700273, 4.3759246, 4.0241704, 3.6285925, 1.989111, 2.23988, 0.56943476, 0.18734744, 3.0697944,
     0.1423234, 4.2426243, 0.74285144, 0.65288776, 2.047052, 1.8705976, 1.7859575, 3.4288344, 2.5029886, 0.9128939,
     3.563948, 4.5813203, 0.26170316, 3.0009906, 2.36652, 3.6201482, 4.392103, 3.4845622, 4.1541553, 1.1975536,
     3.3441992, 4.21749, 2.6703851, 1.0331514, 3.6882434, 2.5319877, 2.2350106, 2.1345248, 1.5940849, 4.4454904,
     1.9453975, 0.35071623, 0.9155592, 1.8320662, 1.4187037, 1.0984321, 1.1173764, 0.9227623, 1.0445648, 2.543863,
     4.3061886, 4.5955496, 4.567751, 0.384735, 1.0590909, 3.5006683, 3.9546578, 4.6020393, 0.45313302, 3.8311226,
     3.904046, 1.9164599, 3.1905115, 2.0310705, 2.7440643, 0.30119884, 2.4038908, 2.6999903, 3.3439562])
x = x.reshape(1, 3 * n_particles, 1)
p1 = [(np.array([1.033667]), np.array([0.])), ()]
params_MM = [p1]
​
dp_x = jax.jit(jax.vmap(jax.vmap(jax.grad(p_x), in_axes=(0, None, None), out_axes=0), \
                        in_axes=(2, None, None), out_axes=2))
​
​
def E_kin(x, p):
    """
    Computes the kinetic energy part of the local energy
    Attributes:
    x: np.ndarray (n_samples, n_dims, n_chains)
        samples from wavefunction
    p: parameters
        Parameters of the Ansatz wavefunction
    Returns:
    ------
    Kinetic energy: np.ndarray (n_samples, n_chains)
    """
​
    unity = np.eye(x.shape[1])
​
    def second_deriv(i, state):
        save, temp = state
        unit = temp[jax.ops.index[i, :]].reshape(1, x.shape[1])
        d2p = jax.jit(jax.vmap(jax.vmap(jax.grad(lambda x, a, b: np.vdot(unit, jax.grad(p_x)(x, a, b))), \
                                        in_axes=(0, None, None), out_axes=0) \
                               , in_axes=(2, None, None), out_axes=2))
        save = jax.ops.index_update(save, jax.ops.index[:, i, :], d2p(x, p, L)[:, i, :])
        return save, temp
​
    save = np.zeros((x.shape))
    save, unity = jax.lax.fori_loop(0, x.shape[1], second_deriv, (save, unity))
    return -0.5 * np.sum((dp_x(x, p, L) / 2) ** 2 + save / 2, axis=1)
​
​
print(E_kin(x, params_MM))

```
gives me with jaxlib==0.1.60 443.02264 whereas with jaxlib==0.1.55 I get 473.73236 also with 64bit enabled..

I will try the vmap, thank you for that suggestion!
